dns/dnsperfgo/main.go

```diff
@@ -0,0 +1,164 @@
+	flag.DurationVar(&config.queryTimeout, "timeout", 5*time.Second, "The timeout for the DNS query.")
+	flag.StringVar(&config.hostnameFile, "inputfile", "", "Path to the file containing hostnames to lookup. Hostnames should be newline-separated.")
+	flag.BoolVar(&config.queryClusterNames, "query-cluster-names", false, "Indicates whether the query names should be the service names in the cluster.")
+	flag.BoolVar(&config.logQueries, "log-queries", false, "Indicates whether each query should be logged.")
+
+	flag.Parse()
+	log.Printf("Starting dnstest with config parameters - %+v", config)
+	config.metricsServer = startMetricsServer(":9153")
+	registerMetrics()
+	run(&config)
+}
+
+func hostnamesFromConfig(config *Config) []string {
+	var hostnamesArr []string
+	if config.hostnameFile != "" {
+		contents, err := ioutil.ReadFile(config.hostnameFile)
+		if err != nil {
+			log.Fatalf("Failed to read input file %q, err - %v, Exiting.", config.hostnameFile, err)
+		}
+		for _, h := range strings.Split(string(contents), "\n") {
+			h = strings.TrimSpace(h)
+			if h != "" {
+				hostnamesArr = append(hostnamesArr, h)
+			}
+		}
+	}
+	// TODO(prameshj) implement reading from k8s API server
+	return hostnamesArr
+}
+
+func run(config *Config) {
+	hostnames := hostnamesFromConfig(config)
+	if len(hostnames) == 0 {
+		log.Fatalf("No hostnames specified, Exiting.")
+	}
+	log.Printf("Got %d hostnames to lookup\n", len(hostnames))
+	qpsSleepDuration := (1 * time.Second) / time.Duration(config.qps)
+	ticker := time.NewTicker(config.testDuration)
+	defer ticker.Stop()
+	// result stores the dns query counts for logging purpose only.
+	result := struct {
+		errorCount   int
+		timeoutCount int
+		totalCount   int
+	}{}
+	for {
+		for _, h := range hostnames {
+			select {
+			case <-config.stopChan:
+				log.Printf("Completed %d queries, %d errors, %d timeouts.\nStop signal received, Exiting.", result.totalCount, result.errorCount, result.timeoutCount)
+				config.metricsServer.Shutdown(context.TODO())
+				return
+			case <-ticker.C:
+				log.Printf("Completed %d queries, %d errors, %d timeouts.\n", result.totalCount, result.errorCount, result.timeoutCount)
+				// Wait for the test to run for testDuration seconds before firing. Without this reset, the test will only run for (testDuration - idleDuration) seconds.
+				ticker.Reset(config.testDuration + config.idleDuration)
+				time.Sleep(config.idleDuration)
+				log.Print("Retarting DNS lookups.")
+			default:
+				break
+			}
+			// Use nsLookup command rather than net.LookupHost because nslookup sends A and AAAA lookups in parallel(with the same source port) in Alpine base image.
+			// Go program sends them with different source ports. The same source port behavior will trigger the DNS race conditions described in https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts.
+			err, timedOut := runQuery(h, config.queryTimeout, nsLookup)
```

Comment:
Does waiting for this query to finish still mean that we run the expected number of queries indicated by the qps?

Reply:
You sre right.. This wait can cause the actual qps to be lower. Addressed by moving it to a goroutine and verifying the timeout case in a local pod.